File: src/main/cpp/com/sandflow/smpte/regxml/dict/MetaDictionary.cpp
/*
* Copyright (c) 2017, Pierre-Anthony Lemieux <pal@palemieux.com>
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*
* * Redistributions of source code must retain the above copyright notice, this
*   list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright notice,
*   this list of conditions and the following disclaimer in the documentation
*   and/or other materials provided with the distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/

#include "MetaDictionary.h"
#include "com/sandflow/smpte/regxml/definitions/PropertyDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/PropertyAliasDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/ClassDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/CharacterTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/EnumerationTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/ExtendibleEnumerationTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/FixedArrayTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/IndirectTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/IntegerTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/OpaqueTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/RecordTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/RenameTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/SetTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/StreamTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/StringTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/StrongReferenceTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/VariableArrayTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/WeakReferenceTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/LensSerialFloatTypeDefinition.h"
#include "com/sandflow/smpte/regxml/definitions/FloatTypeDefinition.h"

namespace rxml {

	class _AddDefinitionVisitor : public DefinitionVisitor {

	public:
		_AddDefinitionVisitor(MetaDictionary &md, EventHandler *ev = &NULL_EVENTHANDLER) : meta_dictionary(md), event_handler(ev) {};

		bool _visit(Definition* def) {

			const AUID defid = MetaDictionary::createNormalizedAUID(def->identification);

			if (meta_dictionary.getDefinition(defid) != NULL) {

				if (event_handler) event_handler->error("Duplicate definition AUID: ", def->identification.to_string(), "");

				return false;

			}

			if (meta_dictionary.getDefinition(def->symbol) != NULL) {

				if (event_handler) event_handler->error("Duplicate definition Symbol: ", def->symbol, "");

				return false;

			}

			meta_dictionary.definitionsByAUID[defid] = def;
			meta_dictionary.definitionsBySymbol[def->symbol] = def;
			meta_dictionary.definitions.push_back(def);

			return true;
		}

		virtual void visit(const ClassDefinition &def) {
			/* Copy and add */

			Definition *def_copy = new ClassDefinition(def);

			if (_visit(def_copy)) {

				/* add to member index */

				if (def.parentClass.is_valid()) {

					const AUID parentauid = MetaDictionary::createNormalizedAUID(def.parentClass.get());

					meta_dictionary.subclassesOf[parentauid].insert(def.identification);
				}

			} else {

				delete def_copy;

			}
		}

		virtual void visit(const PropertyDefinition &def) {

			/* Copy and add */

			Definition *def_copy = new PropertyDefinition(def);

			if (_visit(def_copy)) {

				/* add to member index */

				const AUID parentauid = MetaDictionary::createNormalizedAUID(def.memberOf);

				meta_dictionary.membersOf[parentauid].insert(def.identification);

			} else {

				delete def_copy;

			}


		}


		virtual void visit(const PropertyAliasDefinition & def) {
			const AUID parentauid = MetaDictionary::createNormalizedAUID(def.memberOf);

			meta_dictionary.membersOf[parentauid].insert(def.identification);
		}

		virtual void visit(const EnumerationTypeDefinition &def) {
			Definition *def_copy = new EnumerationTypeDefinition(def);

			if (! _visit(def_copy)) delete def_copy;
		}

		virtual void visit(const CharacterTypeDefinition &def) {
			Definition *def_copy = new CharacterTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const RenameTypeDefinition & def) {
			Definition *def_copy = new RenameTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const RecordTypeDefinition & def) {
			Definition *def_copy = new RecordTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const StringTypeDefinition & def) {
			Definition *def_copy = new StringTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const LensSerialFloatTypeDefinition & def) {
			Definition *def_copy = new LensSerialFloatTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const IntegerTypeDefinition & def) {
			Definition *def_copy = new IntegerTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const StrongReferenceTypeDefinition & def) {
			Definition *def_copy = new StrongReferenceTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const WeakReferenceTypeDefinition & def) {
			Definition *def_copy = new WeakReferenceTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const ExtendibleEnumerationTypeDefinition & def) {
			Definition *def_copy = new ExtendibleEnumerationTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const VariableArrayTypeDefinition & def) {
			Definition *def_copy = new VariableArrayTypeDefinition(def);

			_visit(def_copy);
		}

		virtual void visit(const FixedArrayTypeDefinition & def) {
			Definition *def_copy = new FixedArrayTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const OpaqueTypeDefinition & def) {
			Definition *def_copy = new OpaqueTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const IndirectTypeDefinition & def) {
			Definition *def_copy = new IndirectTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const StreamTypeDefinition & def) {
			Definition *def_copy = new StreamTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const SetTypeDefinition & def) {
			Definition *def_copy = new SetTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

		virtual void visit(const FloatTypeDefinition& def) {
			Definition* def_copy = new FloatTypeDefinition(def);

			if (!_visit(def_copy)) delete def_copy;
		}

	private:
		MetaDictionary &meta_dictionary;
		EventHandler *event_handler;

	};


	MetaDictionary::MetaDictionary(const AUID & pSchemeID, const std::string & pSchemeURI) : schemeID(pSchemeID), schemeURI(pSchemeURI) {
	}

	MetaDictionary::MetaDictionary() {}

	MetaDictionary::~MetaDictionary() {

		for (std::vector<Definition*>::iterator it = definitions.begin(); it != definitions.end(); ++it) {
			delete *it;
		}

	}

	const Definition * MetaDictionary::getDefinition(const AUID & identification) const {

		std::map<AUID, Definition*>::const_iterator it = definitionsByAUID.find(createNormalizedAUID(identification));

		if (it == definitionsByAUID.end()) return NULL;

		return it->second;
	}

	const Definition *  MetaDictionary::getDefinition(const std::string & symbol) const
	{
		if (definitionsBySymbol.find(symbol) == definitionsBySymbol.end()) return NULL;

		return definitionsBySymbol.at(symbol);
	}

    const std::vector<Definition*>& MetaDictionary::getDefinitions() const {
		return this->definitions;
    }

	std::set<AUID> MetaDictionary::getSubclassesOf(const AUID & identification) const {

		const AUID norm_auid = MetaDictionary::createNormalizedAUID(identification);

		if (subclassesOf.find(norm_auid) == subclassesOf.end()) return std::set<AUID>();

		return subclassesOf.at(norm_auid);
	}

	std::set<AUID> MetaDictionary::getMembersOf(const AUID & identification) const {

		const AUID norm_auid = MetaDictionary::createNormalizedAUID(identification);

		if (membersOf.find(norm_auid) == membersOf.end()) return std::set<AUID>();

		return membersOf.at(norm_auid);
	}

	void MetaDictionary::setSchemeID(const AUID & pSchemeID) {
		this->schemeID = pSchemeID;
	}

	const std::string & MetaDictionary::getSchemeURI() const { return this->schemeURI; }

	void MetaDictionary::setSchemeURI(const std::string & pSchemeURI) {
		this->schemeURI = pSchemeURI;
	}

	void MetaDictionary::setDescription(const std::string & desc) {
		this->description = desc;
	}

	void MetaDictionary::addDefinition(const Definition & def, EventHandler * ev)
	{

		_AddDefinitionVisitor ad(*this);

		def.accept(ad);

	}

	UL MetaDictionary::createNormalizedUL(const UL & ul) {
		UL norm_ul = ul;

		/* set version to 0 */

		norm_ul.setValueOctet(7, 0);

		if (norm_ul.isGroup()) {

			/* set byte 6 to 0x7f */
			norm_ul.setValueOctet(5, 0x7f);

		}

		return norm_ul;
	}

	AUID MetaDictionary::createNormalizedAUID(const AUID & auid) {

		if (auid.isUL()) {
			return createNormalizedUL(auid.asUL());
		} else {
			return auid;
		}

	}

	bool MetaDictionary::normalizedAUIDEquals(const AUID & auidA, const AUID & auidB) {

		return createNormalizedAUID(auidA).equals(createNormalizedAUID(auidB));
	}

}

File: src/main/java/com/sandflow/smpte/regxml/dict/MetaDictionary.java
/*
 * Copyright (c) 2014, pal
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.sandflow.smpte.regxml.dict;

import com.sandflow.smpte.regxml.dict.definitions.CharacterTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.ClassDefinition;
import com.sandflow.smpte.regxml.dict.definitions.Definition;
import com.sandflow.smpte.regxml.dict.definitions.EnumerationTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.ExtendibleEnumerationTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.FixedArrayTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.FloatTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.IndirectTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.IntegerTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.LensSerialFloatTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.OpaqueTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.PropertyAliasDefinition;
import com.sandflow.smpte.regxml.dict.definitions.PropertyDefinition;
import com.sandflow.smpte.regxml.dict.definitions.RecordTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.RenameTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.SetTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.StreamTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.StringTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.StrongReferenceTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.VariableArrayTypeDefinition;
import com.sandflow.smpte.regxml.dict.definitions.WeakReferenceTypeDefinition;
import com.sandflow.smpte.regxml.dict.exceptions.DuplicateSymbolException;
import com.sandflow.smpte.regxml.dict.exceptions.IllegalDefinitionException;
import com.sandflow.smpte.util.AUID;
import com.sandflow.smpte.util.UL;
import com.sandflow.smpte.util.UUID;
import com.sandflow.smpte.util.xml.UUIDAdapter;
import java.io.IOException;
import java.io.Reader;
import java.net.URI;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;
import jakarta.xml.bind.JAXBContext;
import jakarta.xml.bind.JAXBException;
import jakarta.xml.bind.Marshaller;
import jakarta.xml.bind.Unmarshaller;
import jakarta.xml.bind.annotation.XmlAccessType;
import jakarta.xml.bind.annotation.XmlAccessorType;
import jakarta.xml.bind.annotation.XmlElement;
import jakarta.xml.bind.annotation.XmlElementWrapper;
import jakarta.xml.bind.annotation.XmlElements;
import jakarta.xml.bind.annotation.XmlRootElement;
import jakarta.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.w3c.dom.Document;

/**
 * A single RegXML Metadictionary as specified in SMPTE ST 2001-1
 */
@XmlRootElement(name = "Extension")
@XmlAccessorType(XmlAccessType.NONE)
public class MetaDictionary implements DefinitionResolver {

    /**
     * XML Schema Namespace for the XML representation of a RegXML Metadictionary 
     */
    public final static String XML_NS = "http://www.smpte-ra.org/schemas/2001-1b/2013/metadict";

    static AUID createNormalizedAUID(AUID auid) {
        if (auid.isUL()) {
            
            return new AUID(createNormalizedUL(auid.asUL()));
        } else {
            return auid;
        }
    }

    static UL createNormalizedUL(UL ul) {
        byte[] value = ul.getValue().clone();
        /* set version to 0 */
        
        value[7] = 0;
        
        if (ul.isGroup()) {
            
            /* set byte 6 to 0x7f */
            value[5] = 0x7f;
            
        }
        
        return new UL(value);
    }

    static String createQualifiedSymbol(String namespace, String symbol) {
        if (namespace == null || namespace.length() == 0) {
            return symbol;
        } else {
            return namespace + " " + symbol;
        }
    }
    
    /**
     * Reads a MetaDictionary from an XML document.
     * 
     * @param reader Reader from which a single MetaDictionary in XML form will be read
     * @return a MetaDictionary
     * @throws JAXBException
     * @throws IOException
     * @throws IllegalDefinitionException 
     */
    public static MetaDictionary fromXML(Reader reader) throws JAXBException, IOException, IllegalDefinitionException {
        JAXBContext ctx = JAXBContext.newInstance(MetaDictionary.class);
        
        Unmarshaller m = ctx.createUnmarshaller();
        MetaDictionary md = (MetaDictionary) m.unmarshal(reader);
        
        for (Definition def : md.definitions) {
            
            def.setNamespace(md.getSchemeURI());
            
            md.indexDefinition(def);
        }

        return md;
    }

    

    @XmlJavaTypeAdapter(value = UUIDAdapter.class)
    @XmlElement(name = "SchemeID", required = true)
    private UUID schemeID;

    @XmlElement(name = "SchemeURI", required = true)
    private URI schemeURI;

    @XmlElement(name = "Description")
    private String description;

    private final ArrayList<Definition> definitions = new ArrayList<>();
    private final HashMap<AUID, Definition> definitionsByAUID = new HashMap<>();
    private final HashMap<String, Definition> definitionsBySymbol = new HashMap<>();
    private final HashMap<AUID, Set<AUID>> membersOf = new HashMap<>();
    private final HashMap<AUID, Set<AUID>> subclassesOf = new HashMap<>();

    private MetaDictionary() {
    }

    /**
     * Instantiates a MetaDictionary.
     * @param schemeURI Scheme URI of the MetaDictionary
     */
    public MetaDictionary(URI schemeURI) {
        /* NOTE: ST 2001-1 does not allow label to be used in multiple enumerations */

        this.schemeID = UUID.fromURIName(schemeURI);
        this.schemeURI = schemeURI;
    }
    
    void indexDefinition(Definition def) throws IllegalDefinitionException {
        AUID defid = createNormalizedAUID(def.getIdentification());
        
        if (def.getClass() != PropertyAliasDefinition.class) {
            
            if (this.definitionsByAUID.containsKey(defid)) {
                throw new IllegalDefinitionException("Duplicate AUID: " + def.getIdentification());
            }
            
            if (this.definitionsBySymbol.containsKey(def.getSymbol())) {
                throw new DuplicateSymbolException("Duplicate Symbol: " + def.getSymbol());
            }
            
            this.definitionsByAUID.put(defid, def);
            this.definitionsBySymbol.put(def.getSymbol(), def);
            
        }
        
        if (def instanceof PropertyDefinition) {
            
            AUID parentauid = createNormalizedAUID(((PropertyDefinition) def).getMemberOf());
            
            Set<AUID> hs = this.membersOf.get(parentauid);
            
            if (hs == null) {
                hs = new HashSet<>();
                this.membersOf.put(parentauid, hs);
            }
            
            hs.add(defid);
            
        }
        
        if (def instanceof ClassDefinition && ((ClassDefinition) def).getParentClass() != null) {
            
            AUID parentauid = createNormalizedAUID(((ClassDefinition) def).getParentClass());
            
            Set<AUID> hs = this.subclassesOf.get(parentauid);
            
            if (hs == null) {
                hs = new HashSet<>();
                this.subclassesOf.put(parentauid, hs);
            }
            
            hs.add(defid);
            
        }
    }

    /**
     * Adds a definition to the MetaDictionary
     * 
     * @param def Definition to be added
     * @throws IllegalDefinitionException 
     */
    public void add(Definition def) throws IllegalDefinitionException {

        if (!def.getNamespace().equals(this.getSchemeURI())) {
            throw new IllegalDefinitionException("Namespace does not match Metadictionary Scheme URI: " + def.getSymbol());
        }

        indexDefinition(def);

        this.definitions.add(def);
    }

    /**
     * @return SchemeID field of the MetaDictionay
     */
    public UUID getSchemeID() {
        return schemeID;
    }

    /**
     * @return SchemeURI field of the MetaDictionay
     */
    public URI getSchemeURI() {
        return schemeURI;
    }

    /**
     * @return Description field of the MetaDictionay
     */
    public String getDescription() {
        return description;
    }

    @XmlElementWrapper(name = "MetaDefinitions")
    @XmlElements(value = {
        @XmlElement(name = "ClassDefinition",
                type = ClassDefinition.class),
        @XmlElement(name = "PropertyDefinition",
                type = PropertyDefinition.class),
        @XmlElement(name = "PropertyAliasDefinition",
                type = PropertyAliasDefinition.class),
        @XmlElement(name = "TypeDefinitionCharacter",
                type = CharacterTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionEnumeration",
                type = EnumerationTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionExtendibleEnumeration",
                type = ExtendibleEnumerationTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionFixedArray",
                type = FixedArrayTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionIndirect",
                type = IndirectTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionInteger",
                type = IntegerTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionOpaque",
                type = OpaqueTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionRecord",
                type = RecordTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionRename",
                type = RenameTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionSet",
                type = SetTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionStream",
                type = StreamTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionString",
                type = StringTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionStrongObjectReference",
                type = StrongReferenceTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionVariableArray",
                type = VariableArrayTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionWeakObjectReference",
                type = WeakReferenceTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionFloat",
                type = FloatTypeDefinition.class),
        @XmlElement(name = "TypeDefinitionLenseSerialFloat",
                type = LensSerialFloatTypeDefinition.class)
    })
    public ArrayList<Definition> getDefinitions() {
        return definitions;
    }

    @Override
    public Definition getDefinition(AUID id) {
        return definitionsByAUID.get(createNormalizedAUID(id));
    }

    /**
     * Retrieves a Definition based on its symbol 
     * @param symbol Symbol of the definition to be retrieved
     * @return Definition, or null if no definition exists with the specified symbol
     */
    public Definition getDefinition(String symbol) {
        return definitionsBySymbol.get(symbol);
    }

    /**
     * Generates an XML representation of the MetaDictionary according to SMPTE ST 2001-1
     * @return The XML DOM
     */
    public Document toXML() {

        Document doc = null;

        try {

            doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Marshaller m = JAXBContext.newInstance(MetaDictionary.class).createMarshaller();
            m.marshal(this, doc);

        } catch (JAXBException | ParserConfigurationException e) {

            throw new RuntimeException(e);

        }

        return doc;
    }

    @Override
    public Collection<AUID> getSubclassesOf(ClassDefinition parent) {

        return subclassesOf.get(createNormalizedAUID(parent.getIdentification()));
    }

    @Override
    public Collection<AUID> getMembersOf(ClassDefinition parent) {
        return membersOf.get(createNormalizedAUID(parent.getIdentification()));
    }


}
